### DISCUSSION TITLE
How to destroy then insert in the same route?

### DISCUSSION THREAD

**Discussion by abcd-ca** [maxcountryman/axum-sessions#57] [Q&A]
Hi, I tried calling `session.destroy()` in my route handler and then `session.insert()` because I want this particular route handler to be able to start with a clean slate whenever it is called. I found that the `insert` call failed. Is it possible to `destroy()` then `insert()` in the same handler one line after the other?

**Comment by maxcountryman** [comment]
Currently this is not possible.

Destroying a session marks it as destroyed. Later the session is removed (at the end of the request). However, at this point your insert is not possible since the session no longer exists and a new session won't be created until the next request.

This life cycle is somewhat fundamental to our design and even more broadly to sessions as a concept. Generally, you can't destroy a session through a reference and then expect that same reference to point to a new session.

So instead of this, what I think would be a better fit for this model is to `clear` the session: this provides a clean slate without interfering with the life cycle semantics of sessions.

**Reply by andrew-cfx** [reply]
Hi, I didn't find a `clear` function so I assume you mean that I should roll my own. This would work fine but it would be cool if it were built in.

**Reply by maxcountryman** [reply]
I think it would make sense to add it to the session API. That said, the session API is not within the direct purview of this crate. So with that in mind, I've actually started work on a new crate that would replace this crate (and I hope) improve it significantly. Stay tuned if that interests you also feedback is appreciated in #56.

**Reply by maxcountryman** [reply]
Here's an example of what that could look like: https://github.com/maxcountryman/tower-sessions/blob/28b49ecdfec57f56f483134dbd218cc6a80c7ca4/src/session.rs#L263-L277.

**Reply by andrew-cfx** [reply]
I like that. I think it's important that the user doesn't need to maintain a list of things to clear that could become out of sync. Your solution appears to manage that well. Maybe internally it can have an initial state that replaces the current state like an Option<T>. T for the session's inner data and None when cleared.

**Reply by maxcountryman** [reply]
We've released [tower-sessions](https://github.com/maxcountryman/tower-sessions) 0.1.0 which includes this functionality. Please let me know how it goes if you give it a try.
